pkg/scheduler/queue/query_component_load.go

```diff
@@ -0,0 +1,187 @@
+// representing whether the query request will be served by the ingesters, store-gateways, or both.
+// Query requests utilizing both ingesters and store-gateways are tracked in the map entry for both keys,
+// therefore the sum of inflight requests by component is likely to exceed the inflight requests total.
+type QueryComponentLoad struct {
+	overloadFactor float64
+
+	inflightRequestsMu                 sync.RWMutex
+	querierInflightRequestsByComponent map[QueryComponent]int
+	querierInflightRequestsTotal       int
+
+	querierInflightRequestsGauge *prometheus.GaugeVec
+}
+
+// QueryComponentDefaultOverloadFactor component is overloaded if it has double the inflight requests as the other
+const QueryComponentDefaultOverloadFactor = 2.0
+
+func NewQueryComponentLoad(overloadFactor float64, registerer prometheus.Registerer) (*QueryComponentLoad, error) {
+	if overloadFactor <= 1 {
+		return nil, errors.New("overloadFactor must be greater than 1")
+	}
+
+	return &QueryComponentLoad{
+		querierInflightRequestsByComponent: make(map[QueryComponent]int),
+		querierInflightRequestsTotal:       0,
+		overloadFactor:                     overloadFactor,
+		querierInflightRequestsGauge: promauto.With(registerer).NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "cortex_query_scheduler_querier_inflight_requests",
+				Help: "Number of inflight requests being processed on a querier-scheduler connection.",
+			},
+			[]string{"query_component"},
+		),
+	}, nil
+}
+
+// IsOverloadedForComponentName checks if load for a query component name is at or above the overload threshold.
+// The component name can indicate the usage of one or more QueryComponents
+// but only one component can be considered to be overloaded at a time,
+// as the threshold is determined by the second-most-loaded-component.
+func (qcl *QueryComponentLoad) IsOverloadedForComponentName(name string) (bool, QueryComponent) {
+	isIngester, isStoreGateway := queryComponentFlags(name)
+
+	qcl.inflightRequestsMu.RLock()
+	defer qcl.inflightRequestsMu.RUnlock()
+
+	overloadThreshold := qcl.overloadThreshold()
+	if overloadThreshold <= 0 {
+		return false, ""
+	}
+
+	if isIngester {
+		if qcl.querierInflightRequestsByComponent[Ingester] >= overloadThreshold {
+			return true, Ingester
+		}
+	}
+	if isStoreGateway {
+		if qcl.querierInflightRequestsByComponent[StoreGateway] >= overloadThreshold {
+			return true, StoreGateway
+		}
+	}
+	return false, ""
+}
+
+// overloadThreshold calculates the minimum number of inflight requests to qualify a query component as overloaded.
+// This is determined by (inflight requests for second-most-loaded component) * (overload factor).
+// Intended to only be called within a read lock on inflightRequestsMu.
+func (qcl *QueryComponentLoad) overloadThreshold() int {
+	// overload factor is expected to be > 1; any misconfiguration of this should be a no-op
+	if qcl.overloadFactor <= 1 {
+		return 0
+	}
```

Comment:
Is this necessary given the validation of `overloadFactor` in `NewQueryComponentLoad`?